onmt/data/Preprocessor.lua

```diff
@@ -1012,18 +1050,29 @@ function Preprocessor:makeData(dataset, dicts)
       local sampledCount = self.args.gsample
       if sampledCount < 1 then
         sampledCount = sampledCount * self.totalCount
+      else
+        sampledCount = sampledCount - self.keepCount
+      end
+      if sampledCount < 0 then
+        _G.logger:error('You requested a sample of %d sentences but %d are already reserved by keep rules. You should configure a larger sample or keep less sentences.',
```

Comment:
works perfectly but I found a problem if all of the files are covered by a keep_rules, and `-gsample N` does not match line count - there is no error message - I propose it mentions it cannot reach the sampling size.

Reply:
Thank. I added a warning in that case.